### PR TITLE
Support granular users

### LINKED ISSUE #10: There should be an option in Postgres CRD to drop database and role on deletion
This could be a variable `spec.keepResources` in the `Postgres` CRD of type boolean.

When set to true it should keep the owner group role and database on deletion. This should be default.

When set to false it should drop the owner group role and database on deletion.

### PR DESCRIPTION
**Pull request by arnarg**
Closes #10 and #18 

I decided to pull the `DropOnDelete` changes into this one as I needed to change it a little bit with this one.

This pull request changes so that `Postgres` can now create schemas in a database. It will maintain reader and writer roles for the schemas it is responsible for that we can grant to `PostgresUser` chosen by the `privileges` property (`OWNER|READ|WRITE`).

- The reader role only has `SELECT` on all tables in all schemas specified in `Postgres`.
- The writer role has `SELECT,INSERT,UPDATE,DELETE` on all tables in all schemas specified in `Postgres`.

You can still skip specifying any schemas in `Postgres` but then you just have to manage readers and writers on your own (or just be owner) as the ones created by the operator will not be aware of the schemas.

I also made sure we don't mutate the spec. I use ArgoCD to maintain our Helm charts and it will be unhappy with any differences in spec.

### CODE REVIEW

**Comment by arnarg** [comment]
@hitman99 The postgres package and the reconcile functions are becoming kind of a mess.

I'd like to refactor them a little bit after this pull request.

I want to split `pkg/postgres/postgres.go` into many smaller files with functions groupped together depending on relation (i.e. `database.go`, `role.go`, `utils.go`, ...).

The reconcile functions are already split into sections (`deletion logic`, `creation logic`, ...) I'd like to have those sections become separate isolated functions (maybe in separate files depending on relation).

Would that be ok?

**Comment by hitman99** [comment]
@arnarg yeah, refactoring is a welcome change! Good call with not mutating the `spec`, I think that at the time I was writing the Status update logic there was no way of updating just the `status`.

**Comment by hitman99** [comment]
@arnarg I'd like to thank you for your contributions and I'm going to make you a maintainer of postgres-operator 😉 But we'll still need to review each other's pull requests before merging 😄

**Comment by hitman99** [comment]
I'm kind of swamped at the moment, but I'll try to review this PR as soon as I can

**Comment by arnarg** [comment]
> I'd like to thank you for your contributions and I'm going to make you a maintainer of postgres-operator

Oh wow! Thanks!

> But we'll still need to review each other's pull requests before merging

Of course! :wink: I also really like discussions in issues and pull requests, it makes solving the problems much easier.

> I'm kind of swamped at the moment, but I'll try to review this PR as soon as I can

No problem

File: pkg/controller/postgres/postgres_controller.go [hunk 1]
```diff
@@ -118,27 +130,89 @@ func (r *ReconcilePostgres) Reconcile(request reconcile.Request) (reconcile.Resu
...
+			continue
+		}
+		err = r.pg.SetSchemaPrivileges(database, owner, writer, schema, writerPrivs, reqLogger)
+		if err != nil {
+			reqLogger.Error(err, fmt.Sprintf("Could not give %s permissions \"%s\"", writer, writerPrivs))
+			continue
+		}
+
+		instance.Status.Schemas = append(instance.Status.Schemas, schema)
+	}
```
[the remark this answers is not in the document]
Comment (arnarg) [reply pkg/controller/postgres/postgres_controller.go]: I agree

**Pull request merged**